codespace: progress indicator, stdlib logger

josebalius opened the pull request:
I've merged the progress indicator into this PR. I think it is ready for review. The major TODO outstanding is figuring out the spinner color.

![progress-indicator](https://user-images.githubusercontent.com/339080/137821291-b2184287-6b01-4df6-80f9-213c9b7768d0.gif)


---
OP:

- Updates the `NewApp` to accept an `iostreams.IOStreams` pointer instead of an `output.Logger`
- `App` now creates an error logger, and checks for interactivity (TTY)
- The application still needs to Print UI-like behavior (for example `.` as a progress indicator) - Ideally this would be done with the spinner, I have another PR that does this but trying to keep the diffs reasonable. In order to support a dotter-like UI, we are using `*App.Print` and `*App.Println` to do so which noops in non-interactive environments, matching the existing behavior.
- I couldn't use a`*log.Logger` for app printing, because `*log.Logger` automatically appends a `\n` to all logs, so it is impossible to use the logger to print to the same line, once we switch to a progress indicator, we can switch to a logger like this since we don't need same-line printing.

This PR should essentially retain the exact same behavior we have today except without `*output.Logger`

josebalius committed:
Spike to change loggers to stdlib

josebalius committed:
A more generic printer to support dot loaders

josebalius committed:
Pass App as logger to Live Share conn creation

josebalius committed:
Use of progress indicator and states

josebalius committed:
Add TODO for choosing functions

josebalius committed:
Switch happy path

josebalius committed:
Remove unnecessary change

josebalius committed:
Fix tests

josebalius committed:
Merge branch 'trunk' of github.com:cli/cli into jg/universe-prep

josebalius committed:
Remove useless logger in favor of App methods

josebalius committed:
Merge branch 'jg/universe-prep' of github.com:cli/cli into jg/use-progress

josebalius committed:
PR Feedback

josebalius committed:
PR Feedback

josebalius committed:
Merge branch 'trunk' of github.com:cli/cli into jg/universe-prep

josebalius committed:
Merge branch 'jg/universe-prep' of github.com:cli/cli into jg/use-progress

josebalius committed:
Use zero value

josebalius committed:
Fix linter

josebalius committed:
Fix race condition

josebalius committed:
Add progress indication to stop command, log ready

josebalius committed:
Merge pull request #4556 from cli/jg/use-progress

codespace: use progress indicator

josebalius committed:
Merge branch 'trunk' of github.com:cli/cli into jg/universe-prep

josebalius commented:
I think this is ready for 👀  - I plan to look at how we can possibility set the right spinner color tomorrow based on terminal bg. I think that's the only thing blocking from this shipping (apart from reviews). Feel free to pull down the PR and run a few commands, it "feels" snappier, I like the spinner.

adonovan commented:
> I plan to look at how we can possibility set the right spinner color 

What's wrong with simply using the foreground color? Presumably the user chose it for its contrast against the background.

josebalius commented:
>What's wrong with simply using the foreground color? Presumably the user chose it for its contrast against the background.

Nothing, I am actually starting to think it's my terminal that's messed up.

1. The `termenv` pkg the CLI uses seems to think I have a dark bg (it's grayish)
2. If I pass `fgBlack` as a color to the spinner, I get the light non-visible color. 

🤔

josebalius committed:
Make spinner fgBlack on light themes

josebalius commented:
I've updated the logic to make it `fgBlack` on light themes instead of using the default white. It is still doesn't for me but I think it's due to my config.

adonovan commented on pkg/iostreams/iostreams.go:
```diff
@@ -229,10 +229,28 @@ func (s *IOStreams) SetNeverPrompt(v bool) {
+	// 11 means Braille. See https://github.com/briandowns/spinner#available-character-sets
+	sp := spinner.New(spinner.CharSets[11], 400*time.Millisecond, opts...)
+
+	if s.TerminalTheme() == "light" {
```
I can't really guess what this does: a "terminal theme" is not a concept in the ANSI terminal spec; it appears to be a heuristic inferred from a handful of nonstandard environment variables such as COLORTERM, NO_COLOR, CLICOLOR, and CLICOLOR_FORCE.

I would hope that saying nothing at all about color here would yield a sensible behavior. Indeed, it does for me, when I run this PR without this line: I see gray braille animated against a white background.

josebalius replied:
It has to be my setup then 😭 - happy to remove this. I think @mislav was going to take a look.

mislav committed:
Improve progress indicator rendering

- Use "cyan" instead of "white" color which is barely legible on light
  terminal backgrounds
- Use 120ms instead of 400ms rendering delay so that the spinner
  feels faster
- Rename StartProgressIndicatorWithSuffix to
  StartProgressIndicatorWithLabel to be agnostic to presentation logic
- Enable changing the label of the currently active progress indicator
- Make Start/StopProgressIndicator methods parallelism-safe

mislav committed:
Cleanup in codespaces printing and logging
